### PR TITLE
Localhost access without SSH

### PR DESCRIPTION
**Pull request by kke**
A host with local address can now be accessed without ssh in all commands.

- The whole `Pharos::SSH` tree was refactored into `Pharos::Transport`.
- Added `pharos exec` alias for `pharos ssh`

72 files changed..

### CODE REVIEW

**Comment by kke** [comment]
I successfully ran `pharos up` on a linux host with one of the hosts being `address: 127.0.0.1` using this.

**Comment by jakolehm** [comment]
@kke travis says no (it runs against localhost).

**Comment by kke** [comment]
image

Hmm yes I suppose it should use sudo for file operations. But I think that's actually easy, it can probably use the same RemoteFile thing but with local-exec transport.

**Comment by kke** [comment]
Travis says yes now.

File: lib/pharos/configuration/host.rb [hunk 1]
```diff
@@ -228,6 +223,8 @@ def configure_bastion(bastion)
 
         attributes[:bastion] = bastion
       end
+
+      def_delegators :transport, :tempfile, :exec, :exec!, :exec?, :exec_script!, :interactive_session, :file, :connected?, :disconnect
```
Comment (jakolehm) [inline lib/pharos/configuration/host.rb]: Maybe move this to top?

File: addons/openebs/addon.rb [hunk 2]
```diff
@@ -43,7 +43,7 @@
 
   reset_host { |host|
     data_dir = config.default_storage_pool[:path].strip
-    host.ssh.exec("sudo rm -rf #{data_dir}/*") unless data_dir.empty?
+    host.exec("sudo rm -rf #{data_dir}/*") unless data_dir.empty?
```
Comment (kke) [inline addons/openebs/addon.rb]: These could actually be `transport.exec` and the host wouldn't need the delegator.

File: lib/pharos/autoload.rb [hunk 3]
```diff
@@ -11,6 +11,8 @@
 autoload :K8s, 'k8s-client'
 autoload :Excon, 'excon'
 autoload :K8s, 'k8s-client'
+autoload :Open3, 'open3'
```
Comment (jakolehm) [inline lib/pharos/autoload.rb]: Just a note that this autoload thing is starting to look a bit monster-ish... we probably should look how to make autoloading better.

File: lib/pharos/configuration/host.rb [hunk 4]
```diff
@@ -78,25 +80,17 @@ def short_hostname
...
+        IPAddr.new(address).loopback?
+      end
 
-        opts = {}
-        opts[:keys] = [ssh_key_path] if ssh_key_path
-        opts[:send_env] = [] # override default to not send LC_* envs
-        opts[:proxy] = Net::SSH::Proxy::Command.new(ssh_proxy_command) if ssh_proxy_command
-        opts[:bastion] = bastion if bastion
-        @ssh = Pharos::SSH::Client.new(address, user, opts.merge(options)).tap(&:connect)
+      def transport(**options)
```
Comment (jakolehm) [inline lib/pharos/configuration/host.rb]: Yardocs would be nice.

File: lib/pharos/phases/authenticate_ssh.rb [hunk 5]
```diff
@@ -6,7 +6,9 @@ class AuthenticateSSH < Pharos::Phase
       title "Authenticate SSH connection"
 
       def call
-        host.ssh(non_interactive: false)
+        return if host.local?
```
Comment (jakolehm) [inline lib/pharos/phases/authenticate_ssh.rb]: Is this necessary? Authentication is skipped if local.

File: lib/pharos/phases/configure_client.rb [hunk 6]
```diff
@@ -18,19 +18,19 @@ def call
         return if @optional && !kubeconfig?
 
         cluster_context['kubeconfig'] = kubeconfig
-        cluster_context['master-ssh'] = ssh
+        cluster_context['master-ssh'] = host.transport
```
Comment (jakolehm) [inline lib/pharos/phases/configure_client.rb]:
```suggestion
        cluster_context['master-transport'] = host.transport
```

Comment (kke) [reply lib/pharos/phases/configure_client.rb]: I dropped the whole thing and replaced with `master_host.transport` where it was used.

**Comment by kke** [comment]
Still need to tweak kubeconfig / ssh subcommands to perform the ssh connection + authentication properly

**Comment by kke** [comment]
Should be PTAL'able again.

I ditched the two-phase ssh auth and made `ssh.connect` do the retry with `non_interactive: false` on its own.

Also added a `pharos kubeconfig` step to drone e2e.

**Pull request merged**